File: ruby/lib/steam/servers/master_server.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'errors/timeout_error'
require 'steam/packets/a2m_get_servers_batch2_packet'
require 'steam/packets/c2m_checkmd5_packet'
require 'steam/packets/s2m_heartbeat2_packet'
require 'steam/servers/server'
require 'steam/sockets/master_server_socket'

# This class represents a Steam master server and can be used to get game
# servers which are publicly available
#
# An intance of this class can be used much like Steam's server browser to get
# a list of available game servers, including filters to narrow down the search
# results.
#
# @author Sebastian Staudt
class MasterServer

  include Server

  # The default number of allowed retries
  @@retries = 3

  # The master server address to query for GoldSrc game servers
  GOLDSRC_MASTER_SERVER = 'hl1master.steampowered.com', 27010

  # The master server address to query for GoldSrc game servers
  SOURCE_MASTER_SERVER = 'hl2master.steampowered.com', 27011

  # The region code for the US east coast
  REGION_US_EAST_COAST = 0x00

  # The region code for the US west coast
  REGION_US_WEST_COAST = 0x01

  # The region code for South America
  REGION_SOUTH_AMERICA = 0x02

  # The region code for Europe
  REGION_EUROPE = 0x03

  # The region code for Asia
  REGION_ASIA = 0x04

  # The region code for Australia
  REGION_AUSTRALIA = 0x05

  # The region code for the Middle East
  REGION_MIDDLE_EAST = 0x06

  # The region code for Africa
  REGION_AFRICA = 0x07

  # The region code for the whole world
  REGION_ALL = 0xFF

  # Sets the number of consecutive requests that may fail, before getting
  # the server list is cancelled (default: 3)
  #
  # @param [Fixnum] The number of allowed retries
  def self.retries=(retries)
    @@retries = retries
  end

  # Request a challenge number from the master server.
  #
  # This is used for further communication with the master server.
  #
  # @note Please note that this is **not** needed for finding servers using
  #       {#servers}.
  # @return [Fixnum] The challenge number from the master server
  # @see #send_heartbeat
  def challenge
    failsafe do
      @socket.send C2M_CHECKMD5_Packet.new
      @socket.reply.challenge
    end
  end

  # Initializes the socket to communicate with the master server
  #
  # @see MasterServerSocket
  def init_socket
    @socket = MasterServerSocket.new @ip_address, @port
  end

  # Returns a list of game server matching the given region and filters
  #
  # Filtering:
  # Instead of filtering the results sent by the master server locally, you
  # should at least use the following filters to narrow down the results sent
  # by the master server.
  #
  # Available filters:
  #
  # * `\type\d`: Request only dedicated servers
  # * `\secure\1`: Request only secure servers
  # * `\gamedir\[mod]`: Request only servers of a specific mod
  # * `\map\[mapname]`: Request only servers running a specific map
  # * `\linux\1`: Request only linux servers
  # * `\emtpy\1`: Request only **non**-empty servers
  # * `\full\`: Request only servers **not** full
  # * `\proxy\1`: Request only spectator proxy servers
  #
  # @note Receiving all servers from the master server is taking quite some
  #       time.
  # @param [Fixnum] region_code The region code to specify a location of the
  #        game servers
  # @param [String] filters The filters that game servers should match
  # @param [Boolean] force Return a list of servers even if an error occured
  #        while fetching them from the master server
  # @raise [SteamCondenser::TimeoutError] if too many timeouts occur while
  #        querying the master server
  # @return [Array<Array<String>>] A list of game servers matching the given
  #         region and filters
  # @see A2M_GET_SERVERS_BATCH2_Packet
  # @see MasterServer.retries=
  def servers(region_code = MasterServer::REGION_ALL, filters = '', force = false)
    fail_count     = 0
    finished       = false
    current_server = '0.0.0.0:0'
    server_array   = []

    begin
      failsafe do
        begin
          @socket.send A2M_GET_SERVERS_BATCH2_Packet.new(region_code, current_server, filters)
          begin
            servers = @socket.reply.servers
            servers.each do |server|
              if server == '0.0.0.0:0'
                finished = true
              else
                current_server = server
                server_array << server.split(':')
              end
            end
            fail_count = 0
          rescue SteamCondenser::TimeoutError
            raise $! if (fail_count += 1) == @@retries
          end
        end while !finished
      end
    rescue SteamCondenser::TimeoutError
      raise $! unless force
    end

    server_array
  end

  # Sends a constructed heartbeat to the master server
  #
  # This can be used to check server versions externally.
  #
  # @param [Hash<Symbol, Object>] The data to send with the heartbeat request
  # @raise [SteamCondenserError] if heartbeat data is missing the
  #        challenge number or the reply cannot be parsed
  # @return [Array<SteamPacket>] Zero or more reply packets from the server.
  #         Zero means either the heartbeat was accepted by the master or there
  #         was a timeout. So usually it's best to repeat a heartbeat a few
  #         times when not receiving any packets.
  # @see S2M_HEARTBEAT2_Packet
  def send_heartbeat(data)
    reply_packets = []

    failsafe do
      @socket.send S2M_HEARTBEAT2_Packet.new(data)

      begin
        loop { reply_packets << @socket.reply }
      rescue SteamCondenser::TimeoutError
      end
    end

    reply_packets
  end

end

File: php/lib/steam/servers/MasterServer.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2011, Sebastian Staudt
 *
 * @author Sebastian Staudt
 * @license http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package Steam Condenser (PHP)
 * @subpackage SteamSocket
 */

require_once STEAM_CONDENSER_PATH . 'InetAddress.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/A2M_GET_SERVERS_BATCH2_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/C2M_CHECKMD5_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/S2M_HEARTBEAT2_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/sockets/MasterServerSocket.php';

/**
 * @package Steam Condenser (PHP)
 * @subpackage MasterServer
 */
class MasterServer
{
	const GOLDSRC_MASTER_SERVER = "hl1master.steampowered.com:27010";
	const SOURCE_MASTER_SERVER = "hl2master.steampowered.com:27011";

        const REGION_US_EAST_COAST = 0x00;
	const REGION_US_WEST_COAST = 0x01;
	const REGION_SOUTH_AMERICA = 0x02;
	const REGION_EUROPE = 0x03;
	const REGION_ASIA = 0x04;
	const REGION_AUSTRALIA = 0x05;
	const REGION_MIDDLE_EAST = 0x06;
	const REGION_AFRICA = 0x07;
	const REGION_ALL = 0xFF;

	/**
	 * @var MasterServerSocket
	 */
	private $socket;

	public function __construct($masterServer)
	{
		$masterServer = explode(":", $masterServer);
		$this->socket = new MasterServerSocket(new InetAddress($masterServer[0]), $masterServer[1]);
	}

    /**
     * Request a challenge number from the master server. This is used for
     * further communication with the master server.
     *
     * Please note that this is NOT needed for finding servers using
     * getServers()
     *
     * @return The challenge number returned from the master server
     * @see sendHeartbeat()
     */
    public function getChallenge() {
        $this->socket->send(new C2M_CHECKMD5_Packet());
        return $this->socket->getReply()->getChallenge();
    }


	public function getServers($regionCode = MasterServer::REGION_ALL , $filter = "")
	{
		$failCount  = 0;
		$finished   = false;
		$portNumber = 0;
		$hostName   = "0.0.0.0";

		do
		{
			$this->socket->send(new A2M_GET_SERVERS_BATCH2_Packet($regionCode, "$hostName:$portNumber", $filter));
			try {
				$serverStringArray = $this->socket->getReply()->getServers();

				foreach($serverStringArray as $serverString)
				{
					$serverString = explode(":", $serverString);
					$hostName = $serverString[0];
					$portNumber = $serverString[1];

					if($hostName != "0.0.0.0" && $portNumber != 0)
					{
						$serverArray[] = array($hostName, $portNumber);
					}
					else
					{
						$finished = true;
					}
				}
				$failCount = 0;
			}
			catch(TimeoutException $e) {
				$failCount ++;
				if($failCount == 4) {
			    		throw $e;
				}
			}
		}
		while(!$finished);

		return $serverArray;
	}

    /**
     * Sends a constructed heartbeat to the master server
     *
     * This can be used to check server versions externally.
     *
     * @param array $data The heartbeat data to send to the master server
     * @return SteamPacket[] The reply from the master server – usually zero or
     *         more packets. Zero means either the heartbeat was accepted by
     *         the master or there was a timeout. So usually it's best to
     *         repeat a heartbeat a few times when not receiving any packets.
     */
    public function sendHeartbeat($data) {
        $this->socket->send(new S2M_HEARTBEAT2_Packet($data));

        $replyPackets = array();
        try {
            do {
                $replyPackets[] = $this->socket->getReply();
            } while(true);
        } catch(TimeoutException $e) {}

        return $replyPackets;
    }

}
?>
